### PR TITLE
Add typing for tmt/steps/execute/__init__.py

### PR DESCRIPTION
**Pull request by janhavlin**

### CODE REVIEW

File: tmt/steps/execute/__init__.py [hunk 1]
```diff
@@ -205,33 +219,38 @@ def base_command(cls, method_class=None, usage=None):
...
-        def execute(context, **kwargs):
+        def execute(context: Any, **kwargs: Any) -> None:
             context.obj.steps.add('execute')
             Execute._save_context(context)
 
         return execute
 
     @classmethod
-    def options(cls, how=None):
+    def options(cls, how: Optional[str] = None) -> List[click.Option]:
```
Comment (janhavlin) [inline tmt/steps/execute/__init__.py:229]: As discussed in [#1319](https://github.com/teemtee/tmt/pull/1319#discussion_r917014429), this should return `List[ClickOptionDecoratorType]` which is being added in #1304. So blocking this PR until #1304 is merged.

File: tmt/steps/execute/__init__.py [hunk 2]
```diff
@@ -166,10 +174,11 @@ def requires(self):
         """
         requires = set()
         for plugin in self.phases(classes=ExecutePlugin):
-            requires.update(plugin.requires())
+            # TODO: remove type: ignore once the self.phases() gets proper type
```
Comment (janhavlin) [inline tmt/steps/execute/__init__.py:177]: Also blocked until #1319 is merged.

File: tmt/steps/execute/__init__.py (hunk 1 above)
Comment (janhavlin) [reply tmt/steps/execute/__init__.py]: Resolved.

File: tmt/steps/execute/__init__.py (hunk 2 above)
Comment (janhavlin) [reply tmt/steps/execute/__init__.py]: Resolved.

**Review by optak — APPROVED** [review]
LGTM

File: tmt/steps/execute/__init__.py [hunk 3]
```diff
@@ -211,33 +223,37 @@ def base_command(
...
             Execute._save_context(context)
 
         return execute
 
     @classmethod
-    def options(cls, how=None):
+    def options(cls, how: Optional[str] = None) -> List[tmt.options.ClickOptionDecoratorType]:
         # Add option to exit after the first test failure
-        options = [click.option(
+        return cast(List[tmt.options.ClickOptionDecoratorType], [click.option(
```
Comment (janhavlin) [inline tmt/steps/execute/__init__.py:235]: We should try to get rid of this cast(). PR https://github.com/teemtee/tmt/pull/1324 has https://github.com/teemtee/tmt/pull/1324#discussion_r926567419 on how to do that, so I'll wait if it gets approved before changing it here also.

File: tmt/steps/execute/__init__.py [hunk 4]
```diff
@@ -42,15 +49,16 @@ class Execute(tmt.steps.Step):
...
+    def __init__(self, plan: "tmt.Plan", data: tmt.steps.StepData) -> None:
         """ Initialize execute step data """
         super().__init__(plan=plan, data=data)
         # List of Result() objects representing test results
-        self._results = []
+        self._results: List[tmt.Result] = []
 
         # List of scripts to install
-        self.scripts = []
+        self.scripts: List[str] = []
```
Comment (psss) [inline tmt/steps/execute/__init__.py:61]: I believe here the `self.scripts` is misplaced, I guess that in 8fcd656c @thrix meant to add this to `ExecutePlugin` not the `Execute` step class.

File: tmt/steps/execute/__init__.py [hunk 5]
```diff
@@ -186,14 +195,16 @@ class ExecutePlugin(tmt.steps.Plugin):
...
-    _supported_methods = []
+    _supported_methods: List[Method] = []
 
     # Common keys for all execute plugins
     _common_keys = ["exit-first"]
 
     # Internal executor is the default implementation
     how = 'tmt'
 
+    scripts: Optional[Tuple['Script', 'Script']] = None
```
Comment (psss) [inline tmt/steps/execute/__init__.py:206]: The number of scripts can vary, so this rather be the following?
```suggestion
    scripts: Tuple['Script', ...] = ()
```

File: tmt/steps/execute/__init__.py [hunk 6]
```diff
@@ -203,6 +214,7 @@ def base_command(
 
         # Prepare general usage message for the step
         if method_class:
+            assert usage is not None
```
Comment (psss) [inline tmt/steps/execute/__init__.py:217]: Should not be needed, `usage` is now mandatory.

File: tmt/steps/execute/__init__.py [hunk 7]
```diff
@@ -211,33 +223,41 @@ def base_command(
         @click.option(
             '-h', '--how', metavar='METHOD',
             help='Use specified method for test execution.')
-        def execute(context, **kwargs):
+        def execute(context: Any, **kwargs: Any) -> None:
```
Comment (psss) [inline tmt/steps/execute/__init__.py:226]: Should be `context: click.Context` instead?

File: tmt/steps/execute/__init__.py [hunk 8]
```diff
@@ -255,26 +276,28 @@ def data_path(self, test, filename=None, full=False, create=False):
...
             self.write(
                 metadata_filename, tmt.utils.dict_to_yaml(test._metadata))
         return tests
 
-    def prepare_scripts(self, guest):
+    def prepare_scripts(self, guest: "tmt.steps.provision.Guest") -> None:
         """
         Prepare additional scripts for testing
         """
+        assert self.scripts
```
Comment (psss) [inline tmt/steps/execute/__init__.py:299]: Should not be needed if we `scripts` is not `Optional` any more.

**Review by psss — CHANGES_REQUESTED** [review]
Thanks for preparing this! Added a couple of comments.

File: tmt/steps/execute/__init__.py (hunk 4 above)
Comment (thrix) [reply tmt/steps/execute/__init__.py:61]: +1, seems this was a mistake

Comment (janhavlin) [reply tmt/steps/execute/__init__.py]: Removed.

File: tmt/steps/execute/__init__.py (hunk 5 above)
Comment (janhavlin) [reply tmt/steps/execute/__init__.py]: Updated, thanks!

File: tmt/steps/execute/__init__.py [hunk 9]
```diff
@@ -322,10 +339,10 @@ def check_beakerlib(self, test):
             return tmt.Result(data, name=test.name, interpret=test.result)
         try:
             result = re.search(
-                'TESTRESULT_RESULT_STRING=(.*)', results).group(1)
+                'TESTRESULT_RESULT_STRING=(.*)', results).group(1)  # type: ignore
```
Comment (thrix) [inline tmt/steps/execute/__init__.py:342]: as discussed in IRC, let's revisit this, we are trying to avoid type ignores

**Review by thrix — CHANGES_REQUESTED** [review]

File: tmt/steps/execute/__init__.py [hunk 10]
```diff
@@ -70,17 +75,18 @@ def load(self, extra_keys=None):
...
         super().save(data)
         results = dict([
             (result.name, result.export()) for result in self.results()])
         self.write('results.yaml', tmt.utils.dict_to_yaml(results))
 
-    def _map_old_methods(self):
+    def _map_old_methods(self) -> None:
         """ Map the old execute methods in a backward-compatible way """
         how = self.data[0]['how']
+        assert isinstance(how, str)
```
Comment (thrix) [inline tmt/steps/execute/__init__.py:89]: I wonder, what is the problem here. I would expect `self.data` to be `tmt.steps.StepData`, which is

```
   71 class StepData(TypedDict, total=False):
   72     """
   73     Step data structure
   74     """
   75     name: str
   76     how: str
   77     order: Optional[int]
```

So why we need to assert it here?

Comment (thrix) [reply tmt/steps/execute/__init__.py]: assert was removed, thanks, resolving

**Review by thrix — APPROVED** [review]

**Comment by thrix** [comment]
@psss looks good to me to be merged, all problems seem to be addressed well.

**Review by psss — APPROVED** [review]
Thanks for all the fixes! Looks good now.

**Pull request merged**